icu4c/source/common/localebuilder.cpp

```diff
@@ -0,0 +1,395 @@
+            if (cmp > 0) {
+                if (!new_attributes.isEmpty()) new_attributes.append('_', status_);
+                new_attributes.append(value_str.data(), status_);
+                inserted = true;
+            }
+        }
+        if (!new_attributes.isEmpty()) new_attributes.append('_', status_);
+        new_attributes.append(start, status_);
+        start += strlen(start) + 1;
+    }
+    if (!inserted) {
+        if (!new_attributes.isEmpty()) new_attributes.append('_', status_);
+        new_attributes.append(value_str.data(), status_);
+    }
+    // Not yet in the attributes, set the attribute.
+    extensions_->setKeywordValue(kAttributeKey, new_attributes.data(), status_);
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::removeUnicodeLocaleAttribute(
+    StringPiece value)
+{
+    CharString value_str(value, status_);
+    if (U_FAILURE(status_)) return *this;
+    char *p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) *p = uprv_tolower(*p);
+    if (!ultag_isUnicodeLocaleAttribute(value_str.data(), value_str.length())) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    if (extensions_ == nullptr) return *this;
+    UErrorCode get_error = U_ZERO_ERROR;
+    CharString attributes;
+    CharStringByteSink sink(&attributes);
+    extensions_->getKeywordValue(kAttributeKey, sink, get_error);
+    // get failure, just return
+    if (U_FAILURE(get_error)) return *this;
+    // Do not have any attributes, just return.
+    if (attributes.length() == 0) return *this;
+
+    p = attributes.data();
+    for (int32_t i = 0; i < attributes.length(); i++, p++) {
+        *p = (*p == '_' || *p == '-') ? '\0' : uprv_tolower(*p);
+    }
+
+    const char* start = attributes.data();
+    const char* end = attributes.data() + attributes.length();
+    CharString new_attributes;
+    bool found = false;
+    while (start < end) {
+        if (uprv_strcmp(start, value_str.data()) == 0) {
+            found = true;
+        } else {
+            if (!new_attributes.isEmpty()) new_attributes.append('_', status_);
+            new_attributes.append(start, status_);
+        }
+        start += strlen(start) + 1;
+    }
+    // Found the value in attributes, set the attribute.
+    if (found) {
+        extensions_->setKeywordValue(kAttributeKey, new_attributes.data(), status_);
+    }
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::clear()
+{
+    status_ = U_ZERO_ERROR;
+    language_[0] = 0;
+    script_[0] = 0;
+    region_[0] = 0;
+    delete variant_;
+    variant_ = nullptr;
+    clearExtensions();
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::clearExtensions()
+{
+    delete extensions_;
+    extensions_ = nullptr;
+    return *this;
+}
+
+Locale LocaleBuilder::build(UErrorCode& error)
+{
+    if (U_SUCCESS(error)) {
+        error = status_;
+        CharString lang_script(language_, error);
```

Comment:
interesting but a little confusing

I would either use a char langScript[14] and just uprv_strcat() or similar,

or else make a single CharString with all of the non-empty lang-script-region-variant and use the single-string Locale constructor.